diesel/src/deserialize.rs

```diff
@@ -255,6 +256,146 @@ where
+///
+/// ```rust
+/// # include!("doctest_setup.rs");
+/// #
+/// #[derive(Queryable, PartialEq, Debug)]
+/// struct User {
+///     id: i32,
+///     name: String,
+/// }
+///
+/// # fn main() {
+/// #     run_test();
+/// # }
+/// #
+/// # fn run_test() -> QueryResult<()> {
+/// #     use schema::users::dsl::*;
+/// #     let connection = establish_connection();
+/// let first_user = users.first(&connection)?;
+/// let expected = User { id: 1, name: "Sean".into() };
+/// assert_eq!(expected, first_user);
+/// #     Ok(())
+/// # }
+/// ```
+///
+/// If we want to do additional work during deserialization, we can use
+/// `deserialize_as` to use a different implementation.
+///
+/// ```rust
+/// # include!("doctest_setup.rs");
+/// #
+/// # use schema::users;
+/// # use diesel::backend::{self, Backend};
+/// # use diesel::deserialize::Queryable;
+/// #
+/// struct LowercaseString(String);
+///
+/// impl Into<String> for LowercaseString {
+///     fn into(self) -> String {
+///         self.0
+///     }
+/// }
+///
+/// impl<DB, ST> Queryable<ST, DB> for LowercaseString
+/// where
+///     DB: Backend,
+///     String: Queryable<ST, DB>,
+/// {
+///     type Row = <String as Queryable<ST, DB>>::Row;
+///
+///     fn build(row: Self::Row) -> Self {
+///         LowercaseString(String::build(row).to_lowercase())
+///     }
+/// }
+///
+/// #[derive(Queryable, PartialEq, Debug)]
+/// struct User {
+///     id: i32,
+///     #[diesel(deserialize_as = "LowercaseString")]
+///     name: String,
+/// }
+///
+/// # fn main() {
+/// #     run_test();
+/// # }
+/// #
+/// # fn run_test() -> QueryResult<()> {
+/// #     use schema::users::dsl::*;
+/// #     let connection = establish_connection();
+/// let first_user = users.first(&connection)?;
+/// let expected = User { id: 1, name: "sean".into() };
+/// assert_eq!(expected, first_user);
+/// #     Ok(())
+/// # }
+/// ```
+///
+/// Alternatively, we can implement the trait for our struct manually.
+///
+/// ```rust
+/// # include!("doctest_setup.rs");
+/// #
+/// use schema::users;
+/// use diesel::deserialize::Queryable;
+///
+/// # /*
+/// type DB = diesel::sqlite::Sqlite;
+/// # */
+///
+/// #[derive(PartialEq, Debug)]
+/// struct User {
+///     id: i32,
+///     name: String,
+/// }
+///
+/// impl Queryable<users::SqlType, DB> for User {
+///     type Row = (i32, String);
+///
+///     fn build(row: Self::Row) -> Self {
+///         User {
+///             id: row.0,
+///             name: row.1.to_lowercase(),
+///         }
+///     }
+/// }
+///
+/// # fn main() {
+/// #     run_test();
+/// # }
+/// #
+/// # fn run_test() -> QueryResult<()> {
+/// #     use schema::users::dsl::*;
+/// #     let connection = establish_connection();
+/// let first_user = users.first(&connection)?;
+/// let expected = User { id: 1, name: "sean".into() };
+/// assert_eq!(expected, first_user);
+/// #     Ok(())
+/// # }
+/// ```
+pub trait TableQueryable {
```

Comment:
I think, the best name will be `SelectResult`, `QueryResult` or even `Selectable`, because as you noted, implementors of this trait not necessary represents a table

Reply:
I also think TableQueryable is not a good name.
I think `Selectable` is better, would there be any conflict (or conflict in the future)?

Reply:
I agree that this should have a better name, `Selectable` is not bad, but I'm not 100% sold to it yet.

Reply:
Changed to `QueryableByColumn`.
If there's some better name, welcome to propose.

Reply:
I don't like that name because it's completely obscure. What do `ByColumn` mean? Looks as if usual `Queryable` doesn't query columns.

Maybe `AutoQueryable`? `Auto` means, that struct fields automatically maps to table columns. Although there may be other, unwanted connotations. `Selectable` in this sense a much better. It is similar to `Queryable`, but different.

One other variant is `Mapped`, that means that struct is mapped to some table or view.

Or even just `TableOrView`.

Reply:
I do not like `QueryableByColumn` either, `Selectable` sounds better for me.

That said: This is a discussion that does not really matter till the actual feature implementation is finished.

Reply:
It is not "Auto", the column name could be specified by `#[column_name = "id"]` while `Mapped` and `TableOrView` seems a little unrelated.
I think `Queryable` do not query columns, they query index. `Queryable` query index, `QueryableByName` query names, and `QueryableByColumn` query columns. (but we would use term "SelectExpression" instead of "Column", so it seems `Selectable` more close to `SelectExpression`)
While I worry about `Selectable` would be confused with `SelectableExpression`. And it might also have more opportunity to conflict with other top level ident.

Reply:
Anyway, I'd like to change the trait name and change `Columns` to `SelectExpression` at the last second, I don't have a refactor tools and it is much easier to have `QueryableByColumn` to be `sed` than `Selectable` :)